### PR TITLE
Add darkproject/kd83a_bfg_edition keyboard

### PR DESCRIPTION
**Pull request by gskygithub**
## Description

Add darkproject/kd83a_bfg_edition keyboard.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/darkproject/kd83a_bfg_edition/config.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define DEVICE_VER 0x0002
+#define VENDOR_ID 0x342D
+#define PRODUCT_ID 0xE392
+#define MANUFACTURER GSKY
+#define PRODUCT KD83A_BFG_Edition
```
Comment (drashna) [inline keyboards/darkproject/kd83a_bfg_edition/config.h:26]: These have been moved to info.json
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/darkproject/kd83a_bfg_edition/config.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+// #define DEVICE_VER 0x0002
+// #define VENDOR_ID 0x342D
+// #define PRODUCT_ID 0xE392
+// #define MANUFACTURER GSKY
+// #define PRODUCT KD83A_BFG_Edition
+
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/config.h:27]:
```suggestion
```

File: keyboards/darkproject/kd83a_bfg_edition/info.json [hunk 3]
```diff
@@ -0,0 +1,110 @@
+{
+    "keyboard_name": "kd83a_bfg_edition",
+    "manufacturer": "kd83a_bfg_edition",
+    "url": "",
+    "maintainer": "Gsky",
+    "usb": {
+        "vid": "0x342D",
+        "pid": "0xE392",
+        "device_version": "0.0.3"
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/info.json:9]:
```suggestion
    "keyboard_name": "KD83A_BFG_Edition",
    "manufacturer": "GSKY",
    "url": "",
    "maintainer": "Gsky",
    "usb": {
        "vid": "0x342D",
        "pid": "0xE392",
        "device_version": "0.0.2"
```

File: keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  KC_GRV,   KC_1,     KC_2,     KC_3,     KC_4,     KC_5,     KC_6,     KC_7,     KC_8,     KC_9,     KC_0,     KC_MINS,  KC_EQL,   KC_BSPC,  RGB_HUI, 
+  KC_TAB,   KC_Q,     TO(Win),  KC_E,     KC_R,     KC_T,     KC_Y,     KC_U,     KC_I,     KC_O,     KC_P,     KC_LBRC,  KC_RBRC,  KC_BSLS,  KC_END,    
+  KC_CAPS,  KC_A,     KC_S,     KC_D,     KC_F,     KC_G,     KC_H,     KC_J,     KC_K,     KC_L,     KC_SCLN,  KC_QUOT,            KC_ENT,   KC_PGUP,    
+  KC_LSFT,  KC_Z,     KC_X,     KC_C,     KC_V,     KC_B,     KC_N,     TO(Mac),  KC_COMM,  KC_DOT,   KC_SLSH,  KC_RSFT,            RGB_VAI,  KC_PGDN,    
+  KC_LCTL,  KC_LALT,  KC_LGUI,                      KC_SPC,                                 KC_RALT,  MO(Winfn),KC_RCTL,  RGB_SPD,  RGB_VAD,  RGB_SPI),
+
+};
+
+void rgb_matrix_indicators_user(void) {
+	if (IS_HOST_LED_ON(USB_LED_CAPS_LOCK)) {
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c:170]:
```suggestion
	if (host_keyboard_led_state().caps_lock) {
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+  KC_ESC,   KC_F1,    KC_F2,    KC_F3,    KC_F4,    KC_F5,    KC_F6,    KC_F7,    KC_F8,    KC_F9,    KC_F10,   KC_F11,   KC_F12,   KC_PSCR,  RGB_MOD,  _______,  _______,   
+  KC_GRV,   KC_1,     KC_2,     KC_3,     KC_4,     KC_5,     KC_6,     KC_7,     KC_8,     KC_9,     KC_0,     KC_MINS,  KC_EQL,   KC_BSPC,  RGB_HUI, 
+  KC_TAB,   KC_Q,     TO(Win),  KC_E,     KC_R,     KC_T,     KC_Y,     KC_U,     KC_I,     KC_O,     KC_P,     KC_LBRC,  KC_RBRC,  KC_BSLS,  KC_END,    
+  KC_CAPS,  KC_A,     KC_S,     KC_D,     KC_F,     KC_G,     KC_H,     KC_J,     KC_K,     KC_L,     KC_SCLN,  KC_QUOT,            KC_ENT,   KC_PGUP,    
+  KC_LSFT,  KC_Z,     KC_X,     KC_C,     KC_V,     KC_B,     KC_N,     TO(Mac),  KC_COMM,  KC_DOT,   KC_SLSH,  KC_RSFT,            RGB_VAI,  KC_PGDN,    
+  KC_LCTL,  KC_LALT,  KC_LGUI,                      KC_SPC,                                 KC_RALT,  MO(Winfn),KC_RCTL,  RGB_SPD,  RGB_VAD,  RGB_SPI),
+
+};
+
+void rgb_matrix_indicators_user(void) {
```
Comment (drashna) [inline keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c:169]: Not to self: https://github.com/qmk/qmk_firmware/pull/18450

**Review by drashna — APPROVED** [review]

**Comment by gksygithub** [comment]
@fauxpark , Resolved requested changes. Could you please help me to take a look. Thanks

**Comment by gksygithub** [comment]
Can someone please take a look at this PR please.

**Review by drashna — APPROVED** [review]

**Comment by gksygithub** [comment]
@fauxpark, please take a look at this PR please. Thanks

File: keyboards/darkproject/kd83a_bfg_edition/config.h [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+#pragma once
+
+#include "config_common.h"
+
+
+/* key matrix size */
+#define MATRIX_ROWS 14
+#define MATRIX_COLS 8
+
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/config.h:25]:
```suggestion
```

File: keyboards/darkproject/kd83a_bfg_edition/info.json [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                { "label": "F7", "x": 7.5, "y": 0 },
+                { "label": "F8", "x": 8.5, "y": 0 },
+                { "label": "F9", "x": 9.75, "y": 0 },
+                { "label": "F10", "x": 10.75, "y": 0 },
+                { "label": "F11", "x": 11.75, "y": 0 },
+                { "label": "F12", "x": 12.75, "y": 0 },
+                { "label": "Printscreen", "x": 14, "y": 0 },
+                { "label": "Del", "x": 15.5, "y": 0 },
+                { "label": "z1", "x": 16.25, "y": 0 },
+                { "label": "z2", "x": 17.25, "y": 0 },
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/info.json:30]:
```suggestion
                { "label": "z2", "x": 17.25, "y": 0 },
```

File: keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+                //rgblight_step();
+            } else {
+                host_consumer_send(0);
+                //rgblight_step();
+            }
+            return false; /* Skip all further processing of this key */
+
+        case KC_LPAD:
+            if (record->event.pressed) {
+                host_consumer_send(0x2A0);
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c:50]:
```suggestion
                host_consumer_send(0x2A0);
```

File: keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+               host_system_send(0x9B);
+               // rgblight_step_reverse();
+            } else {
+               host_system_send(0);
+               //rgblight_step_reverse();
+            }
+            return false; /* Skip all further processing of this key */
+
+        case KC_SPOT:
+            if (record->event.pressed) {
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c:81]:
```suggestion
            if (record->event.pressed) {
```

File: keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+               host_consumer_send(0x221);
+               // rgblight_step_reverse();
+            } else {
+                host_consumer_send(0);
+               //rgblight_step_reverse();
+            }
+            return false; /* Skip all further processing of this key */ 
+
+        case KC_EJ:
+            if (record->event.pressed) {
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c:91]:
```suggestion
            if (record->event.pressed) {
```

File: keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+        case KC_EJ:
+            if (record->event.pressed) {              
+               host_consumer_send(0x0B8);
+               // rgblight_step_reverse();
+            } else {
+                host_consumer_send(0);
+               //rgblight_step_reverse();
+            }
+            return false; /* Skip all further processing of this key */
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/keymaps/via/keymap.c:98]:
```suggestion
            return false; /* Skip all further processing of this key */
```

File: keyboards/darkproject/kd83a_bfg_edition/rules.mk [hunk 12]
```diff
@@ -0,0 +1,23 @@
+# MCU name
+MCU = WB32F3G71
+
+# Bootloader selection
+BOOTLOADER = wb32-dfu
+
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/rules.mk:6]:
```suggestion
```

File: keyboards/darkproject/kd83a_bfg_edition/info.json [hunk 13]
```diff
@@ -0,0 +1,110 @@
+{
+    "keyboard_name": "KD83A_BFG_Edition",
+    "manufacturer": "GSKY",
+    "url": "",
+    "maintainer": "Gsky",
+    "usb": {
+        "vid": "0x342D",
+        "pid": "0xE392",
+        "device_version": "0.0.2"
+    },
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/info.json:10]:
```suggestion
    },
    "processor": "WB32F3G71",
    "bootloader": "wb32-dfu",
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Review by fauxpark — CHANGES_REQUESTED** [review]
.

**Comment by gksygithub** [comment]
@fauxpark I didn't get you, could you please explain in detail what are the changes I need to do in this pull request.

**Comment by fauxpark** [comment]
The suggestions above still need to be accepted.

**Comment by Serega2009bk** [comment]
@gksygithub have a release version of this keyboard, it uses a different mcu WB32FQ95

**Comment by SHVD3x** [comment]
My board also came with WB32FQ95

**Comment by gksygithub** [comment]
@Serega2009bk, [SHVD3x](https://github.com/SHVD3x), yes it is correct MCU (WB32FQ95). updated json file, Thanks.

**Comment by gksygithub** [comment]
@drashna , @fauxpark , please take a look at this PR. Thanks.

File: keyboards/darkproject/kd83a_bfg_edition/config.h [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+#define ENABLE_RGB_MATRIX_SOLID_MULTISPLASH
+
+
+#define RGB_MATRIX_CENTER { 87, 22 }
+
+/* Set debounce time to 5ms */
+#define DEBOUNCE 5
+
+/* Force NKRO on boot up regardless of the setting saved in the EEPROM (uncomment to enable it) */
+// #define FORCE_NKRO
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/config.h:126]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by gksygithub** [comment]
@fauxpark , Resolved requested changes. Please take a look, thanks.

File: keyboards/darkproject/kd83a_bfg_edition/config.h [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "config_common.h"
+
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/config.h:20]:
```suggestion
```

File: keyboards/darkproject/kd83a_bfg_edition/config.h [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+#define LOCKING_RESYNC_ENABLE
+
+#define RGB_DISABLE_WHEN_USB_SUSPENDED
+
+/* External spi flash */
+#define EXTERNAL_FLASH_SPI_SLAVE_SELECT_PIN B14
+#define WEAR_LEVELING_BACKING_SIZE (8 * 1024)
+
+#define LED_CAPS_LOCK_PIN C14
+
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/config.h:52]:
```suggestion
```

File: keyboards/darkproject/kd83a_bfg_edition/info.json [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    "keyboard_name": "KD83A_BFG_Edition",
+    "manufacturer": "GSKY",
+    "url": "",
+    "maintainer": "Gsky",
+    "usb": {
+        "vid": "0x342D",
+        "pid": "0xE392",
+        "device_version": "0.0.3"
+    },
+    "processor": "WB32FQ95",
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/info.json:11]:
```suggestion
    "indicators": {
        "caps_lock": "C14"
    },
    "processor": "WB32FQ95",
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by gksygithub** [comment]
@fauxpark Resolved requested changes, thanks.

**Comment by gksygithub** [comment]
@fauxpark Can you please help to review the changes, thanks.

**Comment by gksygithub** [comment]
Can someone take a look at this PR please..., thanks.

**Comment by gksygithub** [comment]
@fauxpark, please take a look at this PR, thanks.

File: keyboards/darkproject/kd83a_bfg_edition/kd83a_bfg_edition.h [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+//  ESC  F1   F2   F3   F4   F5   F6   F7   F8   F9   F10  F11  F12  Prt      Del    Z1 Z2
+//  `    1    2    3    4    5    6    7    8    9    0    -    =    BSpc     Home 
+//  Tab  Q    W    E    R    T    Y    U    I    O    P    [    ]    \\       End    
+//  Caps A    S    D    F    G    H    J    K    L    ;    '         Enter    PgUp   
+//  Sh_L      Z    X    C    V    B    N    M    ,    .    /     Sh_R   Up    PgDn    
+//  Ct_L    Win_L   Alt_L      SPACE          Alt_R   FN   Ct_R    Left Down  Right  
+
+
+#define LAYOUT( \
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/kd83a_bfg_edition.h:32]: Needs to be moved to info.json.

File: keyboards/darkproject/kd83a_bfg_edition/kd83a_bfg_edition.c [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+    {1, CS10_SW5,  CS11_SW5,  CS12_SW5},  //  82, k37, Right
+   
+    {1, CS10_SW10, CS11_SW10, CS12_SW10}, //  83, Z1
+    {1, CS10_SW11, CS11_SW11, CS12_SW11}, //  84, Z2  
+};
+
+#define __ NO_LED
+//#define RGB_LAYOUT LAYOUT
+
+led_config_t g_led_config = {
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/kd83a_bfg_edition.c:126]: Needs to be moved to info.json.

File: keyboards/darkproject/kd83a_bfg_edition/config.h [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+    { C2 }
+#define ENCODERS_PAD_B \
+    { C3 }
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* Hold ESC on start up to clear EEPROM and boot into bootloader mode */
+#define BOOTMAGIC_LITE_ROW 1
+#define BOOTMAGIC_LITE_COLUMN 3
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/config.h:34]: Needs to be moved to info.json.

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by gskygithub** [comment]
@fauxpark , Could you please take a look at this PR, thanks.

File: keyboards/darkproject/kd83a_bfg_edition/kd83a_bfg_edition.h (hunk 18 above)
Comment (fauxpark) [reply keyboards/darkproject/kd83a_bfg_edition/kd83a_bfg_edition.h:32]: This is still outstanding.

**Review by fauxpark — APPROVED** [review]

**Comment by gskygithub** [comment]
@drashna @fauxpark , still is there any modifications in this PR? please kindly help to review or merge it, thanks.

File: keyboards/darkproject/kd83a_bfg_edition/info.json [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+            "solid_reactive_multicross": true,
+            "solid_reactive_nexus": true,
+            "solid_reactive_multinexus": true,
+            "splash": true,
+            "multisplash": true,
+            "solid_splash": true,
+            "solid_multisplash": true
+        },
+        "center_point": [76, 25],
+        "driver": "AW20216",
```
Comment (fauxpark) [inline keyboards/darkproject/kd83a_bfg_edition/info.json:83]:
```suggestion
        "driver": "aw20216",
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/darkproject/kd83a_bfg_edition/keymaps/default/keymap.c [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+            }
+            return false; /* Skip all further processing of this key */
+
+        case KC_SPOT:
+            if (record->event.pressed) {
+                host_consumer_send(0x221);
+            } else {
+                host_consumer_send(0);
+            }
+            return false; /* Skip all further processing of this key */
```
Comment (drashna) [inline keyboards/darkproject/kd83a_bfg_edition/keymaps/default/keymap.c:55]: This keycode exists as `KC_WWW_SEARCH` already.

Comment (drashna) [reply keyboards/darkproject/kd83a_bfg_edition/keymaps/default/keymap.c:55]:
```suggestion
```

File: keyboards/darkproject/kd83a_bfg_edition/keymaps/default/keymap.c [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+
+        case KC_EJ:
+            if (record->event.pressed) {
+                host_consumer_send(0x0B8);
+            } else {
+                host_consumer_send(0);
+            }
+            return false; /* Skip all further processing of this key */
```
Comment (drashna) [inline keyboards/darkproject/kd83a_bfg_edition/keymaps/default/keymap.c:54]: This exists in core, already, as well.

File: keyboards/darkproject/kd83a_bfg_edition/keymaps/default/keymap.c [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+// Each layer gets a name for readability, which is then used in the keymap matrix below.
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+
+enum my_keycodes {
+    KC_SIRI = QK_USER,
+    KC_DND,
+    KC_EJ,
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
```
Comment (drashna) [inline keyboards/darkproject/kd83a_bfg_edition/keymaps/default/keymap.c:28]: These should be moved to the kb level

File: keyboards/darkproject/kd83a_bfg_edition/kd83a_bfg_edition.c [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+            if (record->event.pressed) {
+                host_consumer_send(0xCF);
+            } else {
+                host_consumer_send(0);
+            }
+            return false; /* Skip all further processing of this key */
+        default:
+            return true; /* Process all other keycodes normally */
+    }
+};
```
Comment (noroadsleft) [inline keyboards/darkproject/kd83a_bfg_edition/kd83a_bfg_edition.c:172]: Defines a default behaviour for encoders that works for keymaps compiled in QMK Configurator. (Configurator doesn't use `ENCODER_MAP`.)

```suggestion
};

#if defined(ENCODER_ENABLE)
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) {
        return false;
    }
    if (clockwise) {
        tap_code(KC_VOLU);
    } else {
        tap_code(KC_VOLD);
    }
    return false;
};
#endif
```

File: keyboards/darkproject/kd83a_bfg_edition/readme.md [hunk 26]
```diff
@@ -0,0 +1,23 @@
+# kd83a_bfg_edition 
+
```
Comment (noroadsleft) [inline keyboards/darkproject/kd83a_bfg_edition/readme.md:2]: Would be good to have a photo of the keyboard and/or PCB here.

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Comment by gskygithub** [comment]
@drashna @fauxpark @lesshonor @noroadsleft, can someone please take a look at this PR please.

**Review by noroadsleft — APPROVED** [review]
Looks good to me.

**Review by drashna — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**